### PR TITLE
Create prod

### PR DESCRIPTION
**Pull request by Dromadon**
Creates production environment

Fixes all S3 buckets with AWS managed encryption

Sets cloudfront distribution with S3 origin, in order to prevent public access to S3

Minor improvements in Makefiles

### CODE REVIEW

File: app/tests/functional/features/steps/authentication.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def given_authenticated_on_cognito(context: Context):
     oauth = OAuth2Session(client=client)
 
     token = oauth.fetch_token(
-        token_url='https://hallebarde-dev.auth.eu-west-1.amazoncognito.com/oauth2/token',
+        token_url='https://dev-filetransfer.auth.eu-west-1.amazoncognito.com/oauth2/token',
```
Comment (Dromadon) [inline app/tests/functional/features/steps/authentication.py:18]: We should get this information from TF output, just like we get the client secrets. In an other issue?

File: app/tests/functional/features/steps/exchange.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 @given('I create an exchange')
 def given_create_an_exchange(context: Context):
-    r = requests.post('https://dev.api.bda.ninja/exchanges',
+    r = requests.post('https://dev.api.filetransfer.beta.octo.tools/exchanges',
```
Comment (Mehdi-H) [inline app/tests/functional/features/steps/exchange.py:10]: Maybe we can put the domain name `https://dev.api.filetransfer.beta.octo.tools` in a variable as we use it in multiple test cases

File: app/tests/responses/test_endpoint_responses.py [hunk 3]
```diff
@@ -35,12 +35,12 @@ def test_get_allowed_origin_should_return_wildcard_when_env_is_dev(self, mock_co
...
+        website_hostname = 'ourwebsite'
         mock_config.ENVIRONMENT = 'prod'
         mock_config.WEBSITE_HOSTNAME = website_hostname
 
         # When
         origin = endpoint_responses.get_allowed_origin()
 
         # Then
-        assert origin == website_hostname
+        assert origin == 'https://'+website_hostname
```
Comment (Mehdi-H) [inline app/tests/responses/test_endpoint_responses.py:46]: Does this test pass if we execute it locally ? (I have some doubt about it with the `https` for local tests)

File: deploy/Makefile [hunk 4]
```diff
@@ -13,23 +13,25 @@ help:
...
 build_infra:
-	source ../conf.sh
 	source ../secret.conf.sh
 	cd environments/${env}
 	terragrunt apply-all --terragrunt-non-interactive --terragrunt-exclude-dir website
 
+destroy_infra:
+	source ../secret.conf.sh
+	cd environments/${env}
+	terragrunt destroy-all --terragrunt-non-interactive
```
Comment (Mehdi-H) [inline deploy/Makefile:28]: I would put an `&&` operator here, what do you think ?

```suggestion
	cd environments/${env} && terragrunt destroy-all --terragrunt-non-interactive
```

File: deploy/modules/website/dns_record.tf [hunk 5]
```diff
@@ -1,7 +1,13 @@
 resource "aws_route53_record" "website_record" {
   zone_id = data.aws_route53_zone.zone.id
   name    = var.env == "prod" ? var.route53_zone_name : "${var.env}.${var.route53_zone_name}"
-  type    = "CNAME"
-  ttl     = "300"
-  records = [aws_cloudfront_distribution.website_distribution.domain_name]
+  type    = "A"
+  //ttl     = "300"
+  //records = [aws_cloudfront_distribution.website_distribution.domain_name]
```
Comment (Mehdi-H) [inline deploy/modules/website/dns_record.tf:6]:
```suggestion
```

File: deploy/modules/website/dns_record.tf [hunk 6]
```diff
@@ -1,7 +1,13 @@
 resource "aws_route53_record" "website_record" {
   zone_id = data.aws_route53_zone.zone.id
   name    = var.env == "prod" ? var.route53_zone_name : "${var.env}.${var.route53_zone_name}"
-  type    = "CNAME"
-  ttl     = "300"
-  records = [aws_cloudfront_distribution.website_distribution.domain_name]
+  type    = "A"
+  //ttl     = "300"
```
Comment (Mehdi-H) [inline deploy/modules/website/dns_record.tf:5]: What is the default TTL ?

File: deploy/modules/website/dns_record.tf [hunk 7]
```diff
@@ -1,7 +1,13 @@
...
+  //records = [aws_cloudfront_distribution.website_distribution.domain_name]
+
+  alias {
+    name                   = aws_cloudfront_distribution.website_distribution.domain_name
+    zone_id                = aws_cloudfront_distribution.website_distribution.hosted_zone_id
+    evaluate_target_health = true
```
Comment (Mehdi-H) [inline deploy/modules/website/dns_record.tf:11]: This makes me think, maybe we should talk about HA of this app some day :d

File: hallebarde-frontend/.env [hunk 8]
```diff
@@ -1,6 +1,6 @@
...
-REACT_APP_OAUTH_DOMAIN="hallebarde-dev.auth.eu-west-1.amazoncognito.com"
-REACT_APP_OAUTH_CALLBACK_URL="https://dev.bda.ninja/callback"
-REACT_APP_OAUTH_SIGNOUT_URL="https://dev.bda.ninja/signout"
-REACT_APP_BACKEND_URL="dev.api.bda.ninja"
\ No newline at end of file
+REACT_APP_USER_POOL_ID="eu-west-1_ChoVjs2Xy"
+REACT_APP_USER_POOL_WEB_CLIENT_ID="78m0k0dgc08hnequo7furn0qb7"
+REACT_APP_OAUTH_DOMAIN="dev-filetransfer.auth.eu-west-1.amazoncognito.com"
+REACT_APP_OAUTH_CALLBACK_URL="http://localhost:3000/callback"
+REACT_APP_OAUTH_SIGNOUT_URL="http://localhost:3000/signout"
```
Comment (Mehdi-H) [inline hallebarde-frontend/.env:5]: Does this variable have an effect on production side ? (I ask because of the `localhost` part), and this .env file seems to be environment agnostic as it is at the root folder

File: hallebarde-frontend/src/components/Download.jsx [hunk 9]
```diff
@@ -25,7 +25,7 @@ export default class Download extends Component {
       })
       .then(({ data }) => {
         this.setState({'isLoading': false})
-        window.open(data)
+        window.open(data, '_self')
```
Comment (Mehdi-H) [inline hallebarde-frontend/src/components/Download.jsx:28]: What does `_self` do ? does it open a new tab ?

**Review by Mehdi-H — APPROVED** [review]

File: app/tests/functional/features/steps/exchange.py (hunk 2 above)
Comment (Dromadon) [reply app/tests/functional/features/steps/exchange.py:10]: Totally. I am wondering if func tests improvements should be in a dedicated PR?

File: app/tests/responses/test_endpoint_responses.py (hunk 3 above)
Comment (Dromadon) [reply app/tests/responses/test_endpoint_responses.py:46]: Actually it does, as this just tests if the function returns the correct string and not the actual HTTPS connection

File: deploy/Makefile (hunk 4 above)
Comment (Dromadon) [reply deploy/Makefile:28]: OK

File: deploy/modules/website/dns_record.tf (hunk 7 above)
Comment (Dromadon) [reply deploy/modules/website/dns_record.tf:11]: Yeah we can, I do not see many things… maybe duplicate the bucket in order to have two distinct  AZ for the static hosting. Lambdas are already HA.

File: hallebarde-frontend/.env (hunk 8 above)
Comment (Dromadon) [reply hallebarde-frontend/.env:5]: actually it is has an effect, and it is not agnostic. This `.env` file is the mecanism I found for building the app before upload to the S3 hosting with the right OAuth information in it (Cognito endpoints, redirect URIs, etc.).

The thing being that because S3 offers only _static_ hosting, we cannot dynamically process these information, and they have to be embedded into the app.

You can see a `configure-environment.sh` file at the root. If you look at the `Makefile`, you will see how it is used, I suppose we can make this more explicit?
Basically, it gets the terraform output in order to fill the `.env` file with the right information, depending on the env you indicate: local, dev, or prod. (Local uses the dev cognito instances, but with redirect URIs pointing to localhost).

File: hallebarde-frontend/src/components/Download.jsx (hunk 9 above)
Comment (Dromadon) [reply hallebarde-frontend/src/components/Download.jsx:28]: Nope, it does the opposite. It opens in the same window, which helps us get rid of the popup authorization problems.

File: deploy/modules/website/dns_record.tf (hunk 6 above)
Comment (Dromadon) [reply deploy/modules/website/dns_record.tf:5]: It is the default TTL for the resource, according to AWS docs and we cannot change it either way https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-choosing-alias-non-alias.html
